### DISCUSSION TITLE
[Feature] log with alarm

### DISCUSSION THREAD

**Discussion by crazychengmm** [apache/skywalking#11521] [Q&A] [answered]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

In application logs, if a keyword appears or a keyword matching a regular expression is found, an alert will be triggered.

![image](https://github.com/apache/skywalking/assets/23517542/929a7ea7-2d07-486f-89a9-30e98d61e073)

According to the official documentation, I found that logs are not supported. If logs are supported, could you please provide some hints or an example. 

Thanks

### Are you willing to submit a pull request to implement this on your own?

- [X] Yes I am willing to submit a pull request on my own!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by wu-sheng** [comment]
Hi, Log should not be used as an alerting rule directly. Instead, you should LAL to extra a metric according to your keyword logic. Then this new metric could be used as an indicator for alarm rules.

**Reply by crazychengmm** [reply]
Thanks for your reply. Could you give me some example to read, I want to get the error message include stack info like this:
```

2023-11-09 10:57:54.048 [TID:3270a7088b594f55ba31afc363cdaf0a.48.16994986728183209] [http-nio-8083-exec-8] ERROR c.c.l.g.s.s.b.s.DcInvocationHandlerFactory$MethodHandlerDecorator -failed to execute the request with request body [BaseRequest(apiName=api.idc.omc.syh.tag.info.get, version=1)] due to 
feign.RetryableException: Connection refused (Connection refused) executing POST http://127.0.0.1:6666/api/invoke/xxxxxxxxx?timestamp=1699498672819&key=e488a12122eb05c33d3c123a2fb819d7
	at feign.FeignException.errorExecuting(FeignException.java:65)
	at feign.SynchronousMethodHandler.executeAndDecode(SynchronousMethodHandler.java:105)
	at feign.SynchronousMethodHandler.invoke(SynchronousMethodHandler.java:77)
	at com.cay.lotuslantern.gateway.sdk.spring.boot.sdk.DcInvocationHandlerFactory$MethodHandlerDecorator.loadAndInvoke(DcInvocationHandlerFactory.java:204)
	at com.cay.lotuslantern.gateway.sdk.spring.boot.sdk.DcInvocationHandlerFactory$MethodHandlerDecorator.invoke(DcInvocationHandlerFactory.java:109)
	at feign.ReflectiveFeign$FeignInvocationHandler.invoke(ReflectiveFeign.java:102)
	at com.sun.proxy.$Proxy143.invoke(Unknown Source
```

{
	"scopeId": 1,
	"scope": "SERVICE",
	"name": "serviceB",
	"id0": "23",
	"id1": "",
    "ruleName": "log_rule",
	"alarmMessage": "alarmMessage 2023-11-09 10:57:54.048 [TID:3270a7088b594f55ba31afc363cdaf0a.48.16994986728183209] [http-nio-8083-exec-8] ERROR c.c.l.g.s.s.b.s.DcInvocationHandlerFactory$MethodHandlerDecorator -failed to execute the request with request body [BaseRequest(apiName=api.idc.omc.syh.tag.info.get, version=1)] due to 
feign.RetryableException: Connection refused (Connection refused) executing POST http://127.0.0.1:6666/api/invoke/xxxxxxxxx?timestamp=1699498672819&key=e488a12122eb05c33d3c123a2fb819d7
	at feign.FeignException.errorExecuting(FeignException.java:65)
	at feign.SynchronousMethodHandler.executeAndDecode(SynchronousMethodHandler.java:105)
	at feign.SynchronousMethodHandler.invoke(SynchronousMethodHandler.java:77)
	at com.cay.lotuslantern.gateway.sdk.spring.boot.sdk.DcInvocationHandlerFactory$MethodHandlerDecorator.loadAndInvoke(DcInvocationHandlerFactory.java:204)
	at com.cay.lotuslantern.gateway.sdk.spring.boot.sdk.DcInvocationHandlerFactory$MethodHandlerDecorator.invoke(DcInvocationHandlerFactory.java:109)
	at feign.ReflectiveFeign$FeignInvocationHandler.invoke(ReflectiveFeign.java:102)
	at com.sun.proxy.$Proxy143.invoke(Unknown Source ",
	"startTime": 1560524171000,
    "tags": [{
        "key": "level",
        "value": "CRITICAL"
    }

I am new in skywalking, I would like to share this with our project team as we are currently facing difficulties in troubleshooting issues.

Thanks

**Reply by wu-sheng** [reply]
I don't have time to write this for you. The LAL docs have included all the important parts. 
https://skywalking.apache.org/docs/main/v9.6.0/en/concepts-and-designs/lal/

**Comment by crazychengmm** [comment]
>>> # Metrics value need to be long, double or int
If I send log extra a metric according, it must be string, could it be working?

Thanks

**Reply by wu-sheng** [reply]
It is a Groovy script, any logic works.

**Comment by crazychengmm** [comment]
Thanks. If i have done this work, i will send it to this communtiy for help new guys.

**Comment by crazychengmm** [comment]
> It is a Groovy script, any logic works.
I have done with lal and mal, get the log metrics like this.
![image](https://github.com/apache/skywalking/assets/23517542/a97de8c8-3436-4969-be7d-ebb0e36ef58c)

```
rules:
  - name: log_count_rule
    dsl: |
      filter {
        text {
          abortOnFailure true
          regexp "(?<timestamp>\\d{4}-\\d{2}-\\d{2} \\d{2}:\\d{2}:\\d{2}.\\d{3}) \\[TID:(?<tid>.+?)] \\[(?<thread>.+?)] (?<level>\\w{4,}) (?<msg>.+)"
        }
        extractor {
          metrics {
            timestamp log.timestamp as Long
            labels level: parsed.level, service: log.service, instance: log.serviceInstance, timestamp: log.timestamp, msg: parsed.msg
            name "log_count"
            value 1
          }
        }
        sink {
          if (parsed.level == "INFO") {
            enforcer { }
          }
          if (parsed.level == "ERROR") {
            enforcer { }
          }
          if (parsed.level == "WARN") {
            enforcer { }
          }
        }
      }
```
          metrics {
            timestamp log.timestamp as Long
            labels level: parsed.level, service: log.service, instance: log.serviceInstance, timestamp: log.timestamp, msg: parsed.msg
            name "log_info"
            value parsed.msg as String
          }
i can not extra msg metrics, it is possible to extra with string ?

**Comment by crazychengmm** [comment]
Skywalking Server: 9.2

If i add metrics with String, blow error come out.
![image](https://github.com/apache/skywalking/assets/23517542/7554d418-baf1-488b-8686-e10dcd1d1379)




```

Caused by: org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:
--
Wed, Nov 15 2023 7:18:19 pm | Script1.groovy: 17: [Static type checking] - Cannot find matching method Script1#value(java.lang.String). Please check if the declared type is correct and if the method exists.
Wed, Nov 15 2023 7:18:19 pm | @ line 17, column 7.
Wed, Nov 15 2023 7:18:19 pm | value parsed.msg as String
Wed, Nov 15 2023 7:18:19 pm | ^
Wed, Nov 15 2023 7:18:19 pm |  
Wed, Nov 15 2023 7:18:19 pm | 1 error
Wed, Nov 15 2023 7:18:19 pm |  
Wed, Nov 15 2023 7:18:19 pm | at org.codehaus.groovy.control.ErrorCollector.failIfErrors(ErrorCollector.java:295) ~[groovy-3.0.8.jar:3.0.8]
Wed, Nov 15 2023 7:18:19 pm | at org.codehaus.groovy.control.CompilationUnit$IPrimaryClassNodeOperation.doPhaseOperation(CompilationUnit.java:984) ~[groovy-3.0.8.jar:3.0.8]
Wed, Nov 15 2023 7:18:19 pm | at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:671) ~[groovy-3.0.8.jar:3.0.8]
Wed, Nov 15 2023 7:18:19 pm | at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:635) ~[groovy-3.0.8.jar:3.0.8]
Wed, Nov 15 2023 7:18:19 pm | at groovy.lang.GroovyClassLoader.doParseClass(GroovyClassLoader.java:389) ~[groovy-3.0.8.jar:3.0.8]
Wed, Nov 15 2023 7:18:19 pm | at groovy.lang.GroovyClassLoader.lambda$parseClass$3(GroovyClassLoader.java:332) ~[groovy-3.0.8.jar:3.0.8]
Wed, Nov 15 2023 7:18:19 pm | at org.codehaus.groovy.runtime.memoize.StampedCommonCache.compute(StampedCommonCache.java:163) ~[groovy-3.0.8.jar:3.0.8]
Wed, Nov 15 2023 7:18:19 pm | at org.codehaus.groovy.runtime.memoize.StampedCommonCache.getAndPut(StampedCommonCache.java:154) ~[groovy-3.0.8.jar:3.0.8]
Wed, Nov 15 2023 7:18:19 pm | at groovy.lang.GroovyClassLoader.parseClass(GroovyClassLoader.java:330) ~[groovy-3.0.8.jar:3.0.8]
Wed, Nov 15 2023 7:18:19 pm | at groovy.lang.GroovyShell.parseClass(GroovyShell.java:526) ~[groovy-3.0.8.jar:3.0.8]
Wed, Nov 15 2023 7:18:19 pm | at groovy.lang.GroovyShell.parse(GroovyShell.java:538) ~[groovy-3.0.8.jar:3.0.8]
Wed, Nov 15 2023 7:18:19 pm | at groovy.lang.GroovyShell.parse(GroovyShell.java:570) ~[groovy-3.0.8.jar:3.0.8]


```

Please give me some hint.
I just want to set alarm on app log.

**Reply by wu-sheng** [reply]
Value is a number, recommended as long. String doesn't work for metric system.

**Reply by crazychengmm** [reply]
Is it possible to add alarm at app log which collect by java agent?
I hava log lal and mal now, how to connect the alarm is the problem.

**Reply by wu-sheng** [reply]
> Is it possible to add alarm at app log which collect by java agent?

For raw logs, no, you can't and you should not. You should extract metrics from the logs, and trigger alarm on those metrics.

**Reply by wu-sheng** [reply, marked as answer]
If those logs are span's logs, then, no, you can't
